File: backend/app/services/ollama_model_utils.py
"""Shared helpers for matching Ollama model names."""
from __future__ import annotations


def _model_aliases(name: str) -> set[str]:
    name = name.strip()
    if not name:
        return set()
    aliases = {name}
    if ":" not in name:
        aliases.add(f"{name}:latest")
    elif name.endswith(":latest"):
        aliases.add(name[: -len(":latest")])
    return aliases


def ollama_model_installed(requested: str, installed: list[str]) -> bool:
    if not requested:
        return False
    req_aliases = _model_aliases(requested)
    return any(req_aliases & _model_aliases(name) for name in installed)


def ollama_models_match(a: str, b: str) -> bool:
    return ollama_model_installed(a, [b])

File: dashboard/src/lib/ollamaModelUtils.ts
/** Match Ollama model names (with optional :latest tag). */
function modelAliases(name: string): Set<string> {
  const trimmed = name.trim();
  if (!trimmed) return new Set();
  const aliases = new Set([trimmed]);
  if (!trimmed.includes(":")) {
    aliases.add(`${trimmed}:latest`);
  } else if (trimmed.endsWith(":latest")) {
    aliases.add(trimmed.slice(0, -":latest".length));
  }
  return aliases;
}

export function ollamaModelInstalled(requested: string, installed: string[]): boolean {
  if (!requested) return false;
  const req = modelAliases(requested);
  return installed.some((name) => {
    for (const alias of modelAliases(name)) {
      if (req.has(alias)) return true;
    }
    return false;
  });
}
